### PR TITLE
Model optimizations and Average Puppi-ML Resolution Calculation

### PR DESCRIPTION
**Pull request by hanhiller**
- removed redundant inputs to model (px and py— they are used in the puppi calculation still)
  - NOTE: Currently these are still fed into the model
- removed redundant/useless pdgID information
  - now neg and pos. pdgIDs map to the same particle type (charge already specified)
  - removed pdgIDs which do not appear in the data set [0,1,2]
- embedding output dimension=2 (was 8)
- fixed bug with puppi resolution calculation (was being scaled by ML response correction)
  - Now puppi resolutions will remain fixed between trainings
- added calculation of the average difference of ML and puppi resolutions (over all bins weighted by the number of events in each bin) (greater average difference = better model)
  - This metric gives us a way to qualitatively compare the resolutions between different models
- added error bars to the METx and METy plots ( these upper and lower errors are the widths of the distribution at points slightly wider and narrower than 1 SD )
- For comparisons sake, here are the average resolution differences for the default (current main) mode: 3 layers (64,32,160) 
  - Root workflow: average xRes difference=5.21, average yRes difference=5.30
  - h5 workflow: average xRes difference=4.98, average yRes difference=4.91
- quantized model trains properly
  - 16,6 bit model out preforms the default model (average xRes dif=5.59, average yRes dif=5.54)
- 2 layers (16-32) seems to train well 
  - root work flow: 5.5 hrs, average xRes dif=3.92, average yRes dif =4.04
  - h5 workflow: ~45min, average xRes dif=4.92, yRes difference=5.20

### CODE REVIEW

File: models.py [hunk 1]
```diff
@@ -47,8 +49,52 @@ def dense_embedding(n_features=6, n_features_cat=2, n_dense_layers=3, activation
...
+        else:
+            inputs.append(input_cat)
+        embedding = Embedding(input_dim=embedding_input_dim[i_emb], output_dim=emb_out_dim, embeddings_initializer=initializers.RandomNormal(mean=0, stddev=0.4/emb_out_dim), name='embedding{}'.format(i_emb))(input_cat)
+        embedding = Reshape((number_of_pupcandis, emb_out_dim))(embedding)
+        embeddings.append(embedding)
+
+    x = Concatenate()([inputs[0]] + [emb for emb in embeddings])
+    
+    for i_dense in range(n_dense_layers):
+        x = QDense(8*2**(n_dense_layers-i_dense), activation = activation_quantizer, kernel_quantizer=logit_quantizer, bias_quantizer=logit_quantizer, kernel_initializer='lecun_uniform')(x)
```
Comment (jmduarte) [inline models.py:77]: Can you split the Dense from the Activation, and re-write it t o be:

```
QDense
BatchNormalization
Activation
```

File: models.py [hunk 2]
```diff
@@ -47,8 +49,52 @@ def dense_embedding(n_features=6, n_features_cat=2, n_dense_layers=3, activation
...
+        x = BatchNormalization(momentum=0.95)(x)
+
+    if t_mode == 0:
+        x = qkeras.qpooling.QGlobalAveragePooling1D(name='pool', quantizer=logit_quantizer)(x)
+        #pool size?
+        outputs = QDense(2, name = 'output', bias_quantizer=logit_quantizer, kernel_quantizer=logit_quantizer, activation=activation_quantizer)(x)
```
Comment (jmduarte) [inline models.py:83]: activation here should probably stay `linear`

File: models.py [hunk 3]
```diff
@@ -47,8 +49,52 @@ def dense_embedding(n_features=6, n_features_cat=2, n_dense_layers=3, activation
...
+        #similar to activation='linear'?
+
+    if t_mode == 1:
+        if with_bias:
+            b = QDense(2, name='met_bias', activation = activation_quantizer, kernel_quantizer=logit_quantizer, bias_quantizer=logit_quantizer, kernel_initializer=initializers.VarianceScaling(scale=0.02))(x)
+            pxpy = Add()([pxpy, b])
+        w = QDense(1, name='met_weight', activation = activation_quantizer, kernel_quantizer=logit_quantizer, bias_quantizer=logit_quantizer, kernel_initializer=initializers.VarianceScaling(scale=0.02))(x)
```
Comment (jmduarte) [inline models.py:90]: Same comment about splitting QDense above

File: models.py [hunk 4]
```diff
@@ -47,8 +49,52 @@ def dense_embedding(n_features=6, n_features_cat=2, n_dense_layers=3, activation
     keras_model.get_layer('met_weight_minus_one').set_weights([np.array([1.]), np.array([-1.]), np.array([0.]), np.array([1.])])
 
     return keras_model
+    
+    
+def dense_embedding_quantized(n_features=6, n_features_cat=2, n_dense_layers=2, number_of_pupcandis=100, embedding_input_dim={0: 13, 1: 3}, emb_out_dim=2, with_bias=True, t_mode = 0, logit_total_bits=7, logit_int_bits=2, activation_total_bits=7, logit_quantizer = 'quantized_bits', activation_quantizer = 'quantized_relu', activation_int_bits=2, alpha=1, use_stochastic_rounding=False):
```
Comment (jmduarte) [inline models.py:54]: Can you propagate the updated model flow of `dense_embedding`?

File: train.py [hunk 5]
```diff
@@ -242,17 +266,13 @@ def main():
...
-    # root file path (input)
-    # output path
-    # mode 0 or 1
     parser = argparse.ArgumentParser()
     parser.add_argument('--dataType', action='store', type=str, required=True, choices=['h5', 'root'], help='designate input file path')
     parser.add_argument('--input', action='store', type=str, required=True, help='designate input file path')
     parser.add_argument('--output', action='store', type=str, required=True, help='designate output file path')
     parser.add_argument('--mode', action='store', type=int, required=True, choices=[0, 1], help='0 for L1MET, 1 for DeepMET')
     parser.add_argument('--epochs', action='store', type=int, required=False, default=100)
+    parser.add_argument('--quantized', action='store_true', required=False, help='flag for quantized model, empty for normal model')
```
Comment (jmduarte) [inline train.py:275]: maybe the user should specify the quantizers here somehow?

File: utils.py [hunk 6]
```diff
@@ -83,44 +88,102 @@ def MakePlots(truth_XY, predict_XY, PUPPI_XY, path_out):
...
     plt.ylabel('RespCorr $\sigma$(METX) [GeV]')
+    plt.title(f'Average $\sigma$(METX) Difference = {round(xRes_avgDif,3)}', fontsize = 22)
     plt.savefig(f"{path_out}resolution_metx.png")
-    plt.close()
+
     plt.figure()
-    plt.hlines(bin_resolY_predict, bin_edges[:-1], bin_edges[1:], colors='r', lw=5,
-           label='Predict', linestyles='solid')
+    plt.hlines(bin_resolY_ML, bin_edges[:-1], bin_edges[1:], colors='r', lw=5,
+               label='Predict', linestyles='solid')
```
Comment (jmduarte) [inline utils.py:150]: 'ML' is probably better than 'Predict'?

File: utils.py [hunk 7]
```diff
@@ -83,44 +88,102 @@ def MakePlots(truth_XY, predict_XY, PUPPI_XY, path_out):
...
     plt.hlines(bin_resolY_PUPPI, bin_edges[:-1], bin_edges[1:], colors='g', lw=5,
-           label='PUPPI', linestyles='solid')
+               label='PUPPI', linestyles='solid')
+    yRes_ML_error = abs([bin_resolY_ML_lower, bin_resolY_ML_upper]-bin_resolY_ML)
+    plt.errorbar(bin_edges[:-1]+12, bin_resolY_ML, yerr= yRes_ML_error, fmt='none', color='r')
```
Comment (jmduarte) [inline utils.py:154]: 12 is a magic number?

File: utils.py [hunk 8]
```diff
@@ -83,44 +88,102 @@ def MakePlots(truth_XY, predict_XY, PUPPI_XY, path_out):
...
+    yRes_PUPPI_error = abs([bin_resolY_PUPPI_lower, bin_resolY_PUPPI_upper]-bin_resolY_PUPPI)
+    plt.errorbar(bin_edges[:-1]+8, bin_resolX_PUPPI, yerr= xRes_PUPPI_error, fmt='none', color='g')
```
Comment (jmduarte) [inline utils.py:156]: 8 is a magic number?

File: utils.py [hunk 9]
```diff
@@ -83,44 +88,102 @@ def MakePlots(truth_XY, predict_XY, PUPPI_XY, path_out):
...
-           label='Predict', linestyles='solid')
+    plt.hlines(bin_resolX_ML, bin_edges[:-1], bin_edges[1:], colors='r', lw=5,
+               label='ML', linestyles='solid')
     plt.hlines(bin_resolX_PUPPI, bin_edges[:-1], bin_edges[1:], colors='g', lw=5,
-           label='PUPPI', linestyles='solid')
+               label='PUPPI', linestyles='solid')
+    xRes_ML_error = abs([bin_resolX_ML_lower, bin_resolX_ML_upper]-bin_resolX_ML)
+    plt.errorbar(bin_edges[:-1]+12, bin_resolX_ML, yerr= xRes_ML_error, fmt='none', color='r')
+    xRes_PUPPI_error = abs([bin_resolX_PUPPI_lower, bin_resolX_PUPPI_upper]-bin_resolX_PUPPI)
+    plt.errorbar(bin_edges[:-1]+8, bin_resolX_PUPPI, yerr= xRes_PUPPI_error, fmt='none', color='g')
```
Comment (jmduarte) [inline utils.py:139]: 8/12 are magic numbers

File: utils.py [hunk 10]
```diff
@@ -83,44 +88,102 @@ def MakePlots(truth_XY, predict_XY, PUPPI_XY, path_out):
...
-    bin_resolX_predict, bin_edges, binnumber = binned_statistic(truth_PtPhi[:,0], truth_XY[:,0] - predict_XY[:,0] * sfs_predict, statistic=resolqt, bins=binnings, range=(0,400))
-    bin_resolY_predict, _, _                 = binned_statistic(truth_PtPhi[:,0], truth_XY[:,1] - predict_XY[:,1] * sfs_predict, statistic=resolqt, bins=binnings, range=(0,400))
-    bin_resolX_PUPPI, _, _                = binned_statistic(truth_PtPhi[:,0], truth_XY[:,0] - PUPPI_XY[:,0] * sfs_predict, statistic=resolqt, bins=binnings, range=(0,400))
-    bin_resolY_PUPPI, _, _                = binned_statistic(truth_PtPhi[:,0], truth_XY[:,1] - PUPPI_XY[:,1] * sfs_predict, statistic=resolqt, bins=binnings, range=(0,400))
+
+    def resolqt_upper(y):
+        return(np.percentile(y,90)-np.percentile(y,10))/2.0
+
+    def resolqt_lower(y):
+        return(np.percentile(y,76)-np.percentile(y,22))/2.0
```
Comment (jmduarte) [inline utils.py:99]: what are upper and lower trying to compute here?

the original function is trying to get the "1 sigma" error band [16%, 84%] / 2 (i.e. +/-1sigma roughly corresponds to the middle 68%)

if you were trying to get the "2 sigma" error band, you could do [2.5%, 97.5%] / 2

but these other values seem somewhat random to me?

you could also combine all of these into 1 function with parameters

File: utils.py [hunk 11]
```diff
@@ -83,44 +88,102 @@ def MakePlots(truth_XY, predict_XY, PUPPI_XY, path_out):
...
+    bin_resolX_PUPPI_upper, _, _ = binned_statistic(true_ptPhi[:,0], trueXY[:,0] - puppiXY[:,0] * sfs_PUPPI, statistic=resolqt_upper, bins=binnings, range=(0,400))
+    bin_resolY_PUPPI_upper, _, _ = binned_statistic(true_ptPhi[:,0], trueXY[:,1] - puppiXY[:,1] * sfs_PUPPI, statistic=resolqt_upper, bins=binnings, range=(0,400))
+
+    bin_resolX_ML_lower, bin_edges, binnumber = binned_statistic(true_ptPhi[:,0], trueXY[:,0] - mlXY[:,0] * sfs_ML, statistic=resolqt_lower, bins=binnings, range=(0,400))
+    bin_resolY_ML_lower, _, _ = binned_statistic(true_ptPhi[:,0], trueXY[:,1] - mlXY[:,1] * sfs_ML, statistic=resolqt_lower, bins=binnings, range=(0,400))
+    bin_resolX_PUPPI_lower, _, _ = binned_statistic(true_ptPhi[:,0], trueXY[:,0] - puppiXY[:,0] * sfs_PUPPI, statistic=resolqt_lower, bins=binnings, range=(0,400))
+    bin_resolY_PUPPI_lower, _, _ = binned_statistic(true_ptPhi[:,0], trueXY[:,1] - puppiXY[:,1] * sfs_PUPPI, statistic=resolqt_lower, bins=binnings, range=(0,400))
+
+    # calculate ML and PUPPI resolutions weighted difference
```
Comment (jmduarte) [inline utils.py:122]: could you instead just get the resolution in a single large bin (rather than this weighted calculation?)

i think these should be equivalent and the other is perhaps more robust?

File: train.py [hunk 12]
```diff
@@ -196,7 +210,17 @@ def trainFrom_h5(args):
...
     # Load training model
-    keras_model = dense_embedding(n_features = n_features_pf, n_features_cat=n_features_pf_cat, n_dense_layers=3, activation='tanh', embedding_input_dim = emb_input_dim, number_of_pupcandis = maxNPF, t_mode = t_mode, with_bias=False)
+    if quantized == True:
+    
+        logit_total_bits = 16
+        logit_int_bits = 6
+        activation_total_bits = 16
+        activation_int_bits = 6
+        
+        keras_model = dense_embedding_quantized(n_features = n_features_pf, emb_out_dim=2, n_features_cat=n_features_pf_cat, n_dense_layers=2, activation_quantizer='quantized_relu',embedding_input_dim = emb_input_dim, number_of_pupcandis = maxNPF, t_mode = t_mode, with_bias=False, logit_quantizer = 'quantized_bits', logit_total_bits=logit_total_bits, logit_int_bits=logit_int_bits, activation_total_bits=activation_total_bits, activation_int_bits=activation_int_bits, alpha=1, use_stochastic_rounding=False)
```
Comment (jmduarte) [inline train.py:220]: break this into multiple lines

**Review by jmduarte — CHANGES_REQUESTED** [review]

File: utils.py (hunk 10 above)
Comment (barnard-lab-uw) [reply utils.py:99]: Yeah, if the +-6 percentile points was random— I tried 2sigma  first, but the error bars were so large, it didn't seem that useful. Ill share a plot with you

File: utils.py (hunk 11 above)
Comment (hanhiller) [reply utils.py:122]: Yep agreed, one bin is probably the way to do this!

File: utils.py (hunk 10 above)
Comment (hanhiller) [reply utils.py:99]: They calculate the width of the distribution at slightly narrow and wider points in the distribution. Is there a better way to calculate error? +-6 from 1 SD is arbitrary— I tried 2 𝜎  at first, but the error bars are pretty large(upper bound is essentially double the resolution). If we do this, then what should I make the lower error bar? Width at half of a sigma? I can share a plot on slack.
For the `resolqt` function, it seems like `binned_statistic` only accepts functions with one parameter as a statistic, so I could use `partial(resolqt, upper=#, lower=#)` but this doesn't seem better than 3 functions.

File: utils.py (hunk 9 above)
Comment (hanhiller) [reply utils.py:139]: +10 will center the error bars on the horizontal lines. +8 +12 puts them close without overlapping

File: models.py [hunk 13]
```diff
@@ -47,8 +53,61 @@ def dense_embedding(n_features=6, n_features_cat=2, n_dense_layers=3, activation
...
+        input_cat = Input(shape=(number_of_pupcandis, 1), name='input_cat{}'.format(i_emb))
+        inputs.append(input_cat)
+        embedding = Embedding(input_dim=embedding_input_dim[i_emb], output_dim=emb_out_dim, embeddings_initializer=initializers.RandomNormal(mean=0, stddev=0.4/emb_out_dim), name='embedding{}'.format(i_emb))(input_cat)
+        embedding = Reshape((number_of_pupcandis, emb_out_dim))(embedding)
+        embeddings.append(embedding)
+
+    x = Concatenate()([inputs_cont, pxpy] + [emb for emb in embeddings])
+    
+    for i_dense in range(n_dense_layers):
+        x = QDense(8*2**(n_dense_layers-i_dense), kernel_quantizer=logit_quantizer, bias_quantizer=logit_quantizer, kernel_initializer='lecun_uniform')(x)
```
Comment (jmduarte) [inline models.py:84]: can you make it the same format with `units` as the other model?

File: train.py [hunk 14]
```diff
@@ -185,8 +196,8 @@ def trainFrom_h5(args):
     Xr = [Xi, Xp] + Xc
 
     indices = np.array([i for i in range(len(Yr))])
-    indices_train, indices_test = train_test_split(indices, test_size=0.2, random_state= 7)
-    indices_train, indices_valid = train_test_split(indices_train, test_size=0.2, random_state=7)
+    indices_train, indices_test = train_test_split(indices, test_size=0.1, random_state= 7)
+    indices_train, indices_valid = train_test_split(indices_train, test_size=1/9, random_state=7)
```
Comment (jmduarte) [inline train.py:200]: maybe 1./9.?

File: utils.py (hunk 9 above)
Comment (jmduarte) [reply utils.py:169]: could you add comment or better yet use variable names to define this:

`binWidth = 10`, etc.

**Review by jmduarte — CHANGES_REQUESTED** [review]

File: models.py [hunk 15]
```diff
@@ -22,8 +26,10 @@ def dense_embedding(n_features=6, n_features_cat=2, n_dense_layers=3, activation
 
     x = Concatenate()([inputs_cont, pxpy] + [emb for emb in embeddings])
 
+    units = [52,52]
```
[the remark this answers is not in the document]
Comment (hanhiller) [reply models.py:29]: Sure, I see what you mean. 52,52 did pretty well. So did 8,52 and in the past so has 16,32 (this is probably what I would choose as the default now.

**Review by jmduarte — CHANGES_REQUESTED** [review]

File: models.py [hunk 16]
```diff
@@ -4,9 +4,21 @@
 import tensorflow as tf
 from tensorflow import slice
 from tensorflow.keras import initializers
+import qkeras
+from qkeras.qlayers import QDense, QActivation
 import numpy as np
 
-def dense_embedding(n_features=6, n_features_cat=2, n_dense_layers=3, activation='relu', number_of_pupcandis=100, embedding_input_dim={0: 13, 1: 3}, emb_out_dim=8, with_bias=True, t_mode = 0):
+units = [16,32]
```
[the remark this answers is not in the document]
Comment (hanhiller) [reply models.py]: My thoughts exactly!
